digdag-core/src/main/java/io/digdag/core/agent/OperatorManager.java

```diff
@@ -177,6 +177,19 @@ protected void runWithHeartbeat(TaskRequest request)
                     }
                     callback.taskFailed(request, agentId, buildExceptionErrorConfig(ex).toConfig(cf));  // no retry
                 }
+                catch (Throwable ex) {
+                    logger.error(
+                            LogMarkers.UNEXPECTED_SERVER_ERROR,
+                            "Task failed with unexpected error: {}", ex.getMessage(), ex);
+
+                    // This block catches OutOfMemoryError and its cause can be either deterministic or non-deterministic.
+                    // So, OperatorManager can't always cancel the failed task.
+                    if (request.isCancelRequested()) {
```

Comment:
In this implementation, only when cancel is requested, task will finish with failure.
If canceling request is not set to the task, the task will continue to run with failure until TTL.
Is my understanding correct ? If so, is there any reason to wait for TTL ?

Reply:
This `catch` block can handle non-deterministic retryable error like OutOfMemoryError caused by temporary lack of overall JVM heap not only deterministic error like OOM caused by instantiation of too huge String. So we can't decide to cancel the failed task immediately. But if the failed task has been already running more than 24 hours, we can cancel it without any concern.

Reply:
Make sense.  👍